File: src/herdr_broker/herdr.py
from __future__ import annotations

import asyncio
import json
import os
import stat
from collections.abc import Callable
from pathlib import Path
from typing import Any
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field, ValidationError


class BrokerError(Exception):
    def __init__(self, code: str, native_code: str | None = None):
        super().__init__(code)
        self.code = code
        self.native_code = native_code


class Pane(BaseModel):
    model_config = ConfigDict(strict=True)
    pane_id: str = Field(min_length=1, max_length=256)
    terminal_id: str = Field(min_length=1, max_length=256)
    workspace_id: str = Field(min_length=1, max_length=256)
    tab_id: str = Field(min_length=1, max_length=256)
    label: str | None = None
    title: str | None = None
    terminal_title_stripped: str | None = None
    cwd: str | None = None
    foreground_cwd: str | None = None
    agent: str | None = None
    agent_status: str


class Tab(BaseModel):
    model_config = ConfigDict(strict=True)
    tab_id: str
    workspace_id: str
    label: str
    number: int
    pane_count: int


class Workspace(BaseModel):
    model_config = ConfigDict(strict=True)
    workspace_id: str
    number: int
    label: str
    focused: bool
    pane_count: int
    tab_count: int
    active_tab_id: str


class Herdr:
    def __init__(self, endpoint: Path, timeout: float = 5):
        self.endpoint = endpoint
        self.timeout = timeout
        self.identity: tuple[int, int] | None = None

    def check_socket(self) -> None:
        try:
            info = self.endpoint.lstat()
            if not stat.S_ISSOCK(info.st_mode) or info.st_uid != os.getuid():
                raise BrokerError("herdr_socket_invalid")
            identity = (info.st_dev, info.st_ino)
            if self.identity is not None and identity != self.identity:
                raise BrokerError("herdr_restarted")
            self.identity = identity
        except OSError as exc:
            raise BrokerError("herdr_unavailable") from exc

    async def request(
        self,
        method: str,
        params: dict[str, Any],
        before_write: Callable[[], None] | None = None,
    ) -> dict[str, Any]:
        self.check_socket()
        writer: asyncio.StreamWriter | None = None
        try:
            async with asyncio.timeout(self.timeout):
                reader, writer = await asyncio.open_unix_connection(str(self.endpoint), limit=4 * 1024 * 1024)
                self.check_socket()
                request_id = str(uuid4())
                payload = json.dumps(
                    {"id": request_id, "method": method, "params": params}, ensure_ascii=False
                )
                if before_write:
                    before_write()
                writer.write(payload.encode() + b"\n")
                await writer.drain()
                line = await reader.readline()
                if not line:
                    raise BrokerError("herdr_disconnected")
                response = json.loads(line)
                if not isinstance(response, dict) or response.get("id") != request_id:
                    raise BrokerError("herdr_invalid_response")
                if "error" in response:
                    error = response["error"]
                    raise BrokerError(
                        "herdr_rejected", error.get("code") if isinstance(error, dict) else None
                    )
                result = response.get("result")
                if not isinstance(result, dict):
                    raise BrokerError("herdr_invalid_response")
                return result
        except TimeoutError as exc:
            raise BrokerError("herdr_timeout") from exc
        except (OSError, ValueError, UnicodeError) as exc:
            raise BrokerError(
                "herdr_invalid_response" if isinstance(exc, ValueError) else "herdr_unavailable"
            ) from exc
        finally:
            if writer:
                writer.close()
                try:
                    await writer.wait_closed()
                except (OSError, asyncio.CancelledError):
                    pass

    async def check(self) -> None:
        result = await self.request("ping", {})
        if result.get("type") != "pong" or result.get("version") != "0.9.0" or result.get("protocol") != 22:
            raise BrokerError("herdr_unsupported")

    async def pane(self, pane_id: str) -> Pane:
        result = await self.request("pane.get", {"pane_id": pane_id})
        try:
            pane = Pane.model_validate(result["pane"])
            if result.get("type") != "pane_info" or pane.pane_id != pane_id:
                raise BrokerError("herdr_invalid_response")
            return pane
        except (KeyError, ValidationError) as exc:
            raise BrokerError("herdr_invalid_response") from exc

    async def panes(self, workspace: str) -> list[Pane]:
        result = await self.request("pane.list", {"workspace_id": workspace})
        try:
            if result.get("type") != "pane_list" or not isinstance(result.get("panes"), list):
                raise BrokerError("herdr_invalid_response")
            panes = [Pane.model_validate(row) for row in result["panes"]]
            if any(p.workspace_id != workspace for p in panes):
                raise BrokerError("herdr_invalid_response")
            return panes
        except ValidationError as exc:
            raise BrokerError("herdr_invalid_response") from exc

    async def tabs(self, workspace: str) -> list[Tab]:
        result = await self.request("tab.list", {"workspace_id": workspace})
        try:
            if result.get("type") != "tab_list" or not isinstance(result.get("tabs"), list):
                raise BrokerError("herdr_invalid_response")
            tabs = [Tab.model_validate(row) for row in result["tabs"]]
            if any(t.workspace_id != workspace for t in tabs):
                raise BrokerError("herdr_invalid_response")
            return tabs
        except ValidationError as exc:
            raise BrokerError("herdr_invalid_response") from exc

    async def workspaces(self) -> list[Workspace]:
        result = await self.request("workspace.list", {})
        try:
            if result.get("type") != "workspace_list" or not isinstance(result.get("workspaces"), list):
                raise BrokerError("herdr_invalid_response")
            return [Workspace.model_validate(row) for row in result["workspaces"]]
        except ValidationError as exc:
            raise BrokerError("herdr_invalid_response") from exc

    async def process_info(self, pane_id: str) -> dict[str, Any]:
        result = await self.request("pane.process_info", {"pane_id": pane_id})
        info = result.get("process_info")
        if (
            result.get("type") != "pane_process_info"
            or not isinstance(info, dict)
            or info.get("pane_id") != pane_id
            or not isinstance(info.get("foreground_processes"), list)
            or any(
                not isinstance(process, dict) or not isinstance(process.get("name"), str)
                for process in info["foreground_processes"]
            )
        ):
            raise BrokerError("herdr_invalid_response")
        return info

File: legacy/typescript/src/herdr.ts
import { createConnection } from 'node:net';
import { randomUUID } from 'node:crypto';
import { lstatSync } from 'node:fs';
import { z } from 'zod';

export class BrokerError extends Error {
  constructor(readonly code: string, readonly nativeCode?: string) { super(code); }
}
const id = z.string().min(1).max(256);
const paneSchema = z.object({
  pane_id: id, terminal_id: id, workspace_id: id, tab_id: id,
  cwd: z.string().max(4096).optional(), foreground_cwd: z.string().max(4096).optional(),
  agent_status: z.string().max(64),
  label: z.string().max(1024).nullable().optional(), title: z.string().max(4096).nullable().optional(), terminal_title_stripped: z.string().max(4096).nullable().optional(),
});
export type Pane = z.infer<typeof paneSchema>;
export interface ConsoleScope { workspace_id: string; tab_id: string; terminals: ReadonlyMap<string, string> }

const processSchema = z.object({ pane_id: id, shell_pid: z.number().int().positive().nullable().default(null), foreground_process_group_id: z.number().int().positive().nullable().default(null),
  foreground_processes: z.array(z.object({ pid: z.number().int().positive(), name: z.string().max(256), argv0: z.string().max(4096).nullable().optional(), argv: z.array(z.string().max(65536)).max(256).nullable().optional() })).max(256).default([]), tty: z.string().max(4096).nullable().optional() });
export type ProcessInfo = z.infer<typeof processSchema>;
interface SendHooks { beforeWrite?: () => void; afterWrite?: () => void; onLateAck?: () => void }

export class Herdr {
  generation = 0;
  private endpointIdentity: string | undefined;
  constructor(private readonly endpoint: string, private readonly verifyAuthority: () => void, private readonly scope?: ConsoleScope) {}
  currentEndpoint() {
    try { const info = lstatSync(this.endpoint); return info.isSocket() && this.endpointIdentity === `${info.dev}:${info.ino}`; }
    catch { return false; }
  }
  requirePane(paneId: string) {
    if (this.scope && !this.scope.terminals.has(paneId)) throw new BrokerError('pane_outside_console');
  }
  private requireTarget(pane: Pane) {
    this.requirePane(pane.pane_id);
    if (this.scope && (pane.workspace_id !== this.scope.workspace_id || pane.tab_id !== this.scope.tab_id || this.scope.terminals.get(pane.pane_id) !== pane.terminal_id)) throw new BrokerError('pane_outside_console');
  }
  private request(method: string, params: object, signal?: AbortSignal, submission?: SendHooks): Promise<unknown> {
    return new Promise((resolve, reject) => {
      try {
        this.verifyAuthority();
        const info = lstatSync(this.endpoint);
        if (!info.isSocket()) throw new BrokerError('herdr_unavailable');
        const identity = `${info.dev}:${info.ino}`;
        if (this.endpointIdentity !== undefined && this.endpointIdentity !== identity) this.generation++;
        this.endpointIdentity = identity;
      } catch (error) {
        this.generation++;
        reject(error instanceof BrokerError ? error : new BrokerError('herdr_unavailable')); return;
      }
      const requestId = randomUUID();
      const socket = createConnection(this.endpoint);
      let buffer = Buffer.alloc(0);
      let settled = false;
      let closed = false;
      let deadline: ReturnType<typeof setTimeout> | undefined;
      const finish = (error?: BrokerError, result?: unknown) => {
        if (closed) return;
        closed = true;
        if (error && !submission && !signal?.aborted && ['herdr_timeout', 'herdr_disconnected', 'herdr_unavailable', 'herdr_invalid_response', 'herdr_response_too_large'].includes(error.code)) this.generation++;
        clearTimeout(deadline);
        buffer = Buffer.alloc(0);
        signal?.removeEventListener('abort', abort);
        socket.destroy();
        if (!settled) { settled = true; if (error) reject(error); else resolve(result); }
        else if (!error && z.object({ type: z.literal('ok') }).safeParse(result).success) submission?.onLateAck?.();
      };
      const abort = () => finish(new BrokerError('cancelled'));
      signal?.addEventListener('abort', abort, { once: true });
      if (signal?.aborted) { abort(); return; }
      deadline = setTimeout(() => {
        if (!settled && submission?.onLateAck) {
          settled = true; reject(new BrokerError('herdr_timeout'));
        } else finish(new BrokerError('herdr_timeout'));
      }, 5000);
      socket.on('error', () => finish(new BrokerError('herdr_unavailable')));
      socket.on('end', () => finish(new BrokerError('herdr_disconnected')));
      socket.on('connect', () => {
        try { this.verifyAuthority(); submission?.beforeWrite?.(); socket.write(JSON.stringify({ id: requestId, method, params }) + '\n', () => {
          try { submission?.afterWrite?.(); } catch { finish(new BrokerError('send_interrupted')); }
        }); }
        catch (error) { finish(error instanceof BrokerError ? error : new BrokerError('authority_lost')); }
      });
      socket.on('data', chunk => {
        if (buffer.length + chunk.length > 1024 * 1024) { finish(new BrokerError('herdr_response_too_large')); return; }
        buffer = Buffer.concat([buffer, chunk]);
        const end = buffer.indexOf(10);
        if (end < 0) return;
        try {
          const response = z.object({ id: z.literal(requestId), result: z.unknown().optional(), error: z.object({ code: z.string() }).optional() })
            .parse(JSON.parse(new TextDecoder('utf-8', { fatal: true }).decode(buffer.subarray(0, end))));
          if (response.error) throw new BrokerError(submission && ['pane_not_found', 'invalid_key', 'pane_send_failed'].includes(response.error.code) ? response.error.code : 'herdr_rejected', response.error.code);
          if (!response.result) throw new BrokerError('herdr_invalid_response');
          this.verifyAuthority();
          finish(undefined, response.result);
        } catch (error) { finish(error instanceof BrokerError ? error : new BrokerError('herdr_invalid_response')); }
      });
    });
  }
  async send(paneId: string, payload: { text: string; keys: string[] }, signal: AbortSignal, hooks: SendHooks) {
    this.requirePane(paneId);
    const response = await this.request('pane.send_input', { pane_id: paneId, ...payload }, signal, hooks);
    if (!z.object({ type: z.literal('ok') }).safeParse(response).success) throw new BrokerError('herdr_invalid_response');
  }
  async processInfo(paneId: string, signal?: AbortSignal) {
    this.requirePane(paneId);
    const parsed = z.object({ type: z.literal('pane_process_info'), process_info: processSchema }).safeParse(await this.request('pane.process_info', { pane_id: paneId }, signal));
    if (!parsed.success) throw new BrokerError('herdr_invalid_response');
    if (parsed.data.process_info.pane_id !== paneId) throw new BrokerError('target_changed');
    return parsed.data.process_info;
  }
  async capture(pane: Pane, signal?: AbortSignal, source: 'recent' | 'recent_unwrapped' = 'recent') {
    this.requireTarget(pane);
    const parsed = z.object({ type: z.literal('pane_read'), read: z.object({
      pane_id: id, workspace_id: id, tab_id: id, text: z.string().refine(text => text.isWellFormed()),
      source: z.literal(source), format: z.literal('ansi'), truncated: z.boolean(), revision: z.number().int().nonnegative(),
    }) }).safeParse(await this.request('pane.read', { pane_id: pane.pane_id, source, lines: 1000, format: 'ansi', strip_ansi: false }, signal));
    if (!parsed.success) throw new BrokerError('herdr_invalid_response');
    const read = parsed.data.read;
    if (read.pane_id !== pane.pane_id || read.workspace_id !== pane.workspace_id || read.tab_id !== pane.tab_id) throw new BrokerError('target_changed');
    return read;
  }
  async check(signal?: AbortSignal) {
    const result = await this.request('ping', {}, signal);
    if (!z.object({ type: z.literal('pong'), version: z.literal('0.9.0'), protocol: z.literal(22) }).safeParse(result).success) {
      this.generation++;
      throw new BrokerError('herdr_unsupported');
    }
  }
  async describe(paneId: string, signal?: AbortSignal): Promise<Pane> {
    this.requirePane(paneId);
    await this.check(signal);
    const result = z.object({ type: z.literal('pane_info'), pane: paneSchema }).safeParse(await this.request('pane.get', { pane_id: paneId }, signal));
    if (!result.success) throw new BrokerError('herdr_invalid_response');
    if (result.data.pane.pane_id !== paneId) throw new BrokerError('target_changed');
    this.requireTarget(result.data.pane);
    return result.data.pane;
  }
  async splitPane(source: Pane, cwd: string, direction: 'right' | 'down', ratio = 0.5) {
    const result = z.object({ pane: paneSchema }).safeParse(await this.request('pane.split', { workspace_id: source.workspace_id, target_pane_id: source.pane_id, direction, ratio, cwd, focus: false }));
    if (!result.success || result.data.pane.workspace_id !== source.workspace_id || result.data.pane.tab_id !== source.tab_id) throw new BrokerError('herdr_invalid_response');
    return result.data.pane;
  }
  async list(workspaceId: string) {
    await this.check();
    const result = z.object({ type: z.literal('pane_list'), panes: z.array(paneSchema).max(4096) }).safeParse(await this.request('pane.list', { workspace_id: workspaceId }));
    if (!result.success || result.data.panes.some(pane => pane.workspace_id !== workspaceId)) throw new BrokerError('herdr_invalid_response');
    return result.data.panes;
  }
  async renamePane(paneId: string, label: string) { await this.request('pane.rename', { pane_id: paneId, label }); }
  async closePane(paneId: string) { await this.request('pane.close', { pane_id: paneId }); }
  async notify(title: string, body: string) { await this.request('notification.show', { title, body, sound: 'none' }); }
}
